src/ui/hook/useCopyToClipboard.ts

```diff
@@ -0,0 +1,52 @@
+import { useCallback, useState } from 'react'
+import { useTimeout } from '@/ui/hook/useTimeout'
+
+type CopyStatus = 'idle' | 'success' | 'error'
+
+type Status<T extends CopyStatus> = { status: T }
+type Message = { message: string }
+
+type IdleState = Status<'idle'>
+type SuccessState = Status<'success'> & Message
+type ErrorState = Status<'error'> & Message
+
+type CopyState = IdleState | SuccessState | ErrorState
+type HandleCopy = (textToCopy: string) => Promise<void>
+
+export const useCopyToClipboard = (
+  copyResetDelay?: number
+): { copyState: CopyState; handleCopy: HandleCopy } => {
+  const [copyState, setCopyState] = useState<CopyState>({
+    status: 'idle'
+  })
+
+  const resetCopyState = useCallback((): void => {
+    setCopyState({
+      status: 'idle'
+    })
+  }, [])
+
+  const handleCopy = useCallback(
+    async (textToCopy: string): Promise<void> =>
+      navigator.clipboard
+        .writeText(textToCopy)
+        .then(() => {
+          setCopyState({
+            status: 'success',
+            message: textToCopy
+          })
+        })
+        .catch((error: Error) => {
+          setCopyState({
+            status: 'error',
+            message: error.message
```

Comment:
imho, it'd be better to return the error, rather than taking on the responsibility of returning the message. This way, the caller has the opportunity to interpret the error appropriately, without degrading the original information.